### DISCUSSION TITLE
Assistance With Writing to N64Console

### DISCUSSION THREAD

**Discussion by porkya** [NicoHood/Nintendo#44] [Q&A]
I am attempting to make a simple device that can swap buttons sent from my Nintendo 64 controller to my Nintendo 64. I am powering the Arduino using the 3.3v supply that powers the controller and the console. Both grounds are connected to the same ground on the Arduino, and the controller and console are connected to separate data pins on the Arduino.

I am able to run the example N64Controller code and the controller inputs are received via serial fine, but I cannot write any inputs to the Nintendo 64 console; the console does not receive any controller inputs at all. The sketch below simply reports N64 buttons as normal and attempts to write them to console. I have made a similar device for the GameCube and the sketch used contains code that is the same as what runs on my GameCube except modified to run on the N64:

```
#include "Nintendo.h"

//Initialize N64 Controller
CN64Controller N64Controller(4);

//Initialize N64 Console
CN64Console N64Console(5);

// Pin definitions
#define pinLed LED_BUILTIN

void setup()
{
  // Set up debug led
  pinMode(pinLed, OUTPUT);

}

void loop()
{
    //Just stops the code if no controller is found
    if (!N64Controller.read()) {
      delay(100);
      return;
    }


    // Get the data of controller
    auto r = N64Controller.getReport();
    N64_Data_t d = defaultN64Data;

    //Reporting all controller inputs

    d.report.a = r.a;
    d.report.b = r.b;
    d.report.z = r.z;
    d.report.start = r.start;
    d.report.cup = r.cup;
    d.report.cright = r.cright;
    d.report.cdown = r.cdown;
    d.report.cleft = r.cleft;
    d.report.dleft = r.dleft;
    d.report.dright = r.dright;
    d.report.ddown = r.ddown;
    d.report.dup = r.dup;
    d.report.z = r.z;
    d.report.r = r.r;
    d.report.l = r.l;
    d.report.xAxis = r.xAxis;
    d.report.yAxis = r.yAxis;
    
    // Sends the data to the console 
    if (!N64Console.write(d))
    {
      delay(100);
    }

  }
```
Additionally, I created another program that mimicked the GamecubeConsole sketch, and that gave the error message 'Failed to write to console'. Any help would be greatly appreciated! Thank you

**Comment by Skuzee** [comment]
Not sure what your physical hardware situation looks like. Are you using a logic level converter? are you sure you're using the correct output pin and the correct data pin to the console? 

```
    // Sends the data to the console 
    if (!N64Console.write(d))
    {
      delay(100);
    }
```
Not sure if that can just be simplified to `N64Console.write(d)` just in case that is causing some sort of timing issue.

**Reply by Skuzee** [reply]
The only other thing I can think of is try.
 `d.status = = N64Controller.getStatus();`
 That will make sure that the data you're sending to the console has the correct status that matches the actual controller. The status contains the device id. though default n64 data should be enough?

**Reply by porkya** [reply]
Hi, thanks for the response. I originally did not have a bidirectional logic converter connected, but I added one to the circuit and the same problem persists; I can read from the controller, but I cannot write to the console. My circuit is now as follows:

5v supply ---> logic converter 5v--> Arduino 5v 
3.3v controller and console supply ---> logic converter 3.3v
Controller data ---> logic converter low, logic high ---> Arduino D4
Console data ---> logic converter low, logic high ---> Arduino D5
All grounds are connected together.

I tried eliminating the if statement for the write command, with no success.

As for your most recent advice, did you mean 'd.status = N64Controller.getStatus();' ?I cannot use the comparison operator without a compilation error. Either way, I tried using the command above and still nothing.

**Reply by Skuzee** [reply]
Sorry, that's a typo.
So, you have both data wires connected to Arduino D4 (via llc)? Though your program says pins 4 and 5, so that might be a typo! 
If both data wires end up at pin 4 that's the problem.

I would also note that the n64 console write function was added recently by a different author, so there might be a bug? I have successfully read a n64 controller and converted it and written to a wii. (but that uses the gamecube write.)

**Reply by Skuzee** [reply]
@JonnyHaystack maybe can help. They wrote the n64 write code. (thanks in advance.)

**Reply by porkya** [reply]
Yeah, I typo'd too, haha. I meant that the console is connected to pin 5. I've reconnected and checked my wiring a few times now and have gotten the same result, being able to read but not write.

I looked over some of the source code before posting, particularly the functions to use for N64Controller and N64Console. When compared to the GameCube variant, the code read practically the same. As for the other stuff, I am unable to understand much of it.

Thanks for the tag, hopefully this all can be resolved!

**Reply by JonnyHaystack** [reply]
Not really sure what the problem would be here. The code is basically the same, it just has a slightly different packet structure for the controller inputs and has simpler commands for polling etc. It's been tested by myself and other people I know using the [B0XX](https://b0xx.com/) hardware, and I can't think why it wouldn't work for you unless your wiring is somehow wrong. I can't really comment on that too much because I'm not very familiar with level shifters. Maybe it could be an issue with using that separate power supply instead of the power from the console? Also, maybe try with different games because they poll at different rates (usually the same polling rate as the frame rate I believe).

[This ](https://github.com/JonnyHaystack/HayB0XX/blob/master/HayB0XX/N64Backend.cpp) is my code for wrapping around CN64Console.
I'm not really doing anything special, except that the controller input packet is prepared in advance and then I delay until shortly before when the next poll is expected in order to reduce input lag. Not sure if your delay(100) could be causing any issues.
